mysys_ssl/my_aes.cc

```diff
@@ -57,3 +57,97 @@ void my_aes_create_key(const unsigned char *key, uint key_length,
     *ptr^= *sptr;
   }
 }
+
+int my_aes_encrypt_nopad(const unsigned char *source, uint32 source_length,
+                         unsigned char *dest,
+                         const unsigned char *key, uint32 key_length,
+                         enum my_aes_opmode mode, const unsigned char *iv)
+{
+  unsigned char eiv[MY_AES_BLOCK_SIZE];
+  uint32 num_blocks= (source_length / MY_AES_BLOCK_SIZE) * MY_AES_BLOCK_SIZE;
+  enum my_aes_opmode ecb_mode;
+  int dst_len;
+
+  if (key_length == 64)
+    ecb_mode= my_aes_256_ecb;
+  else if (key_length == 32)
+    ecb_mode= my_aes_128_ecb;
+  else
+    return MY_AES_BAD_DATA;
+
+  my_aes_encrypt(iv, MY_AES_BLOCK_SIZE, eiv,
+                 key, key_length, ecb_mode, NULL, false);
+
+  dst_len= my_aes_encrypt(source, num_blocks, dest,
+                          key, key_length, mode, eiv, false);
+  if (dst_len == MY_AES_BAD_DATA)
+    return dst_len;
+
+  if (num_blocks < source_length)
+  {
+    unsigned char mask[MY_AES_BLOCK_SIZE];
+    uint32 len= source_length - num_blocks;
+
+    DBUG_ASSERT(dst_len == static_cast<int>(num_blocks));
+
+    dst_len= my_aes_encrypt(eiv, MY_AES_BLOCK_SIZE, mask,
+                            key, key_length, ecb_mode, NULL, false);
+
+    if (dst_len == MY_AES_BAD_DATA)
+      return dst_len;
+
+    for (uint32 i= 0; i < len; i++)
```

Comment:
Is this "xoring with encrypted eiv" a well-known approach or a clever workaround? If yes, please give a reference.

Reply:
It actually very similar to how CTR mode works. It takes counter, encrypt it and XOR with plaintext, which allows to parallelize encryption.